code/modules/projectiles/guns/beam_rifle.dm

```diff
@@ -0,0 +1,418 @@
+	..()
+	START_PROCESSING(SSfastprocess, src)
+
+/obj/item/weapon/gun/energy/beam_rifle/Destroy()
+	STOP_PROCESSING(SSfastprocess, src)
+	..()
+
+/obj/item/weapon/gun/energy/beam_rifle/zoom(user, forced_zoom)
+	. = ..()
+	scope(user, .)
+
+/obj/item/weapon/gun/energy/beam_rifle/proc/scope(mob/user, forced)
+	var/scoping
+	switch(forced)
+		if(TRUE)
+			scoping = TRUE
+		if(FALSE)
+			scoping = FALSE
+		else
+			scoping = !scoped
+	if(scoping)
+		spread = scoped_inaccuracy
+		recoil = scoped_recoil
+		scoped = TRUE
+		user << "<span class='boldnotice'>You bring your [src] up and use its scope...</span>"
+	else
+		spread = hipfire_inaccuracy
+		recoil = hipfire_recoil
+		scoped = FALSE
+		user << "<span class='boldnotice'>You lower your [src].</span>"
+
+/obj/item/weapon/gun/energy/beam_rifle/can_trigger_gun(var/mob/living/user)
+	if(!scoped && !noscope)
+		user << "<span class='userdanger'>This beam rifle can only be used while scoped!</span>"
+		return FALSE
+	. = ..(user)
+
+/obj/item/weapon/gun/energy/beam_rifle/emp_act(severity)
+	chambered = null
+	recharge_newshot()
+	return	//Energy drain handled by its cell.
+
+/obj/item/weapon/gun/energy/beam_rifle/proc/aiming_beam()
+	var/atom/A = current_user.client.mouseObject
+	if(!istype(A) || !A.loc)
+		return
+	var/turf/T = get_turf(current_user.client.mouseObject)
+	if(!istype(T))
+		return
+	var/obj/item/projectile/beam/beam_rifle/hitscan/aiming_beam/P = new
+	P.gun = src
+	P.preparePixelProjectile(current_user.client.mouseObject, T, current_user, current_user.client.mouseParams, 0)
+	if(aiming_time >= 1)
+		var/percent = ((100/aiming_time)*aiming_time_left)
+		P.color = rgb(255 * percent,255 * ((100 - percent) / 100),0)
+	else
+		P.color = rgb(0, 255, 0)
+	clear_tracers()
+	P.fire()
+
+/obj/item/weapon/gun/energy/beam_rifle/proc/clear_tracers()
+	for(var/I in current_tracers)
+		current_tracers -= I
+		var/obj/effect/temp_visual/projectile_beam/PB = I
+		PB.alpha = 0
+		qdel(PB)
+
+/obj/item/weapon/gun/energy/beam_rifle/process()
+	if(!aiming)
+		return
+	if(aiming_time_left > 0)
+		aiming_time_left--
+	aiming_beam()
+	if(current_user.client.mouseParams)
+		var/list/mouse_control = params2list(current_user.client.mouseParams)
+		if(mouse_control["screen-loc"])
+			var/list/screen_loc_params = splittext(mouse_control["screen-loc"], ",")
+			var/list/screen_loc_X = splittext(screen_loc_params[1],":")
+			var/list/screen_loc_Y = splittext(screen_loc_params[2],":")
+			var/x = (text2num(screen_loc_X[1]) * 32 + text2num(screen_loc_X[2]) - 32)
+			var/y = (text2num(screen_loc_Y[1]) * 32 + text2num(screen_loc_Y[2]) - 32)
+			var/screenview = (current_user.client.view * 2 + 1) * world.icon_size //Refer to http://www.byond.com/docs/ref/info.html#/client/var/view for mad maths
+			var/ox = round(screenview/2) - current_user.client.pixel_x //"origin" x
+			var/oy = round(screenview/2) - current_user.client.pixel_y //"origin" y
+			var/angle = Atan2(y - oy, x - ox)
+			while(angle > 360)	//Don't round!
```

Comment:
No it won't.
it turns things like 540 to 180 (which is identical)
360 -> 0 (identical)
etc, it clamps angles to 0->359, which is what these while loops do, but the loops do it poorly.